File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution
{
    public:
        int lengthOfLongestSubstring(string s)
        {
            int n = s.length();
            int st = 0, end = 0, ans = 0;
            int freq[128]={};

            while (end < n)
            {
                freq[s[end]]++;
                while (freq[s[end]] > 1)
                {
                    freq[s[st]]--;
                    st++;
                }

                ans = max(ans, end - st + 1);
                end++;
            }

            return ans;
        }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character,Integer>mp = new HashMap<>();
        int n = s.length();
        int st = 0, end = 0;
        int maxLength = 0;

        while(end < n){
            char ch = s.charAt(end);
            mp.put(ch,mp.getOrDefault(ch,0)+1);
            while(mp.get(ch) > 1){
                char curr = s.charAt(st);
                mp.put(curr,mp.get(curr)-1);
                st++;
            }
            maxLength = Math.max(maxLength, end - st + 1);
            end++;

        }
        return maxLength;
    }
}